### PR TITLE
Fix Object#freeze also freezing part of the class

### PR DESCRIPTION
**Pull request by casperisfine**
[[Bug #19164]](https://bugs.ruby-lang.org/issues/19164)

The bug was introduced in 8a13a2e8d1361bdab5d4bc1d75cf07b804a982c0 / https://github.com/ruby/ruby/pull/5955

NB: This is just the failing case for now.

### CODE REVIEW

File: vm_method.c [hunk 1]
```diff
@@ -1541,8 +1541,8 @@ remove_method(VALUE klass, ID mid)
     rb_method_entry_t *me = 0;
     VALUE self = klass;
 
-    klass = RCLASS_ORIGIN(klass);
     rb_class_modify_check(klass);
+    klass = RCLASS_ORIGIN(klass);
```
Comment (casperisfine) [inline vm_method.c:1545]: I'm really unsure wether this fix is correct.

**Comment by XrXr** [comment]
Ah, there is FL_SINGLETON is FL_USER0, so there is false aliasing:
https://github.com/ruby/ruby/blob/01790de9e625b48272dc8ba3e9ce94787fb5a519/include/ruby/internal/fl_type.h#L430
https://github.com/ruby/ruby/blob/e7b1ff984fde859a7778dec564731eb79392406f/internal/class.h#L125

Should add a comment when changing.

**Comment by casperisfine** [comment]
@XrXr does that mean we're using the same flag on the same type for two different things?

If so, I assume the proper fix is to move either flag?

**Comment by XrXr** [comment]
Moving one of them would restore logic to before e7b1ff984fde859a7778dec564731eb79392406f, so might be more conservative than the current patch. Theoretically, it should be fine for these flags to overlap, because `RUBY_FL_SINGLETON` shouldn't be checked on ICLASSes. The code is just setup to not distinguish between iclasses/modules/classes very well.

The current diff is probably fine. It actually seems to fix an unrelated bug:

```
class A
  prepend Module.new
  def foo = nil
  freeze

  remove_method :foo # remove works even though class is frozen!
end
```

**Comment by XrXr** [comment]
For shapes folks @jemmaissroff, @tenderlove: I merged the fix because I think the fix is uncontroversial. There might still be other issues arising from RUBY_FL_SINGLETON sharing the same bit as RICLASS_IS_ORIGIN, though. 8a13a2e8d1361bdab5d4bc1d75cf07b804a982c0 made it so that `Object.new.freeze` also make `RCLASS_ORIGIN(Object)` frozen. Now, we don't use the frozen bit on iclasses now, but maybe this new frozen setting behavior should also be changed. I don't really understand why `rb_freeze_singleton_class()` tries to freeze the origin in the first place. https://github.com/ruby/ruby/blob/3d272b0fc822f5c2e9c716377b7fcecf15426b27/class.c#L2139-L2150

**Pull request merged**